norse/torch/functional/stdp.py

```diff
@@ -1,400 +1,276 @@
-
-    return dw, STDPState(x=x_decayed, y=y_decayed)
-
-
-def conv2d_soft_multiplicative_stdp_step(
-    w: torch.Tensor,
-    z_pre: torch.Tensor,
-    z_post: torch.Tensor,
-    state: STDPState,
-    p: STDPConvParameters = STDPConvParameters(),
-    dt: float = 0.001,
-) -> Tuple[torch.Tensor, STDPState]:
-    """Soft-bound multiplicative STDP step for nn.Conv2d layer, meaning change in
-    synaptic weight is dependent on current synaptic weight. Based on online implementation
-    from http://www.scholarpedia.org/article/Spike-timing_dependent_plasticity
-
-    Parameters:
-        w (torch.Tensor): synaptic weight
-        z_pre (torch.Tensor): presynaptic spikes
-        z_post (torch.Tensor): postsynaptic spikes
-        state (STDPState): STDP state
-        p (STDPConvParameters): STDP and convolutional parameters
-        dt (float): integration time step
-    """
-    dx = dt * p.tau_pre_inv * (-state.x + p.a_pre * z_pre)
-    x_decayed = state.x + dx
-
-    dy = dt * p.tau_post_inv * (-state.y + p.a_post * z_post)
-    y_decayed = state.y + dy
-
-    # Get shape information
-    batch_size = z_pre.shape[0]
-    out_channels, _, kernel_h, kernel_w = w.shape
-
-    # Unfold presynaptic neurons to get receptive field per postsynaptic neuron
-    x_decayed_uf = torch.nn.functional.unfold(
-        x_decayed,
-        (kernel_h, kernel_w),
-        dilation=p.dilation,
-        padding=p.padding,
-        stride=p.stride,
+    # Update post-synaptic layer
+    z_post, state_post = lif_feed_forward_step(
+        torch.nn.functional.linear(z_pre, w),  # Integrate into LIF FF class
+        state_post,
+        p_post,
+        dt,
     )
-    z_pre_uf = torch.nn.functional.unfold(
+
+    # Update STDP traces
+    state_stdp.decay(
         z_pre,
-        (kernel_h, kernel_w),
-        dilation=p.dilation,
-        padding=p.padding,
-        stride=p.stride,
+        z_post,
+        p_stdp.tau_pre_inv,
+        p_stdp.tau_post_inv,
+        p_stdp.a_pre,
+        p_stdp.a_post,
+        dt,
     )
 
-    # Sum before reshape
-    dw_plus = _A_plus_soft(w, p.w_max, p.eta_plus) * torch.bmm(
-        z_post.view(batch_size, out_channels, -1), x_decayed_uf.permute((0, 2, 1))
-    ).sum(0).view(w.shape)
-    dw_minus = _A_minus_soft(w, p.w_min, p.eta_minus) * torch.bmm(
-        y_decayed.view(batch_size, out_channels, -1), z_pre_uf.permute((0, 2, 1))
-    ).sum(0).view(w.shape)
+    # STDP weight update
+    dw_plus = p_stdp.A_plus(w) * torch.einsum("bi,bj->ij", z_post, state_stdp.t_pre)
+    dw_minus = p_stdp.A_minus(w) * torch.einsum("bi,bj->ij", state_stdp.t_post, z_pre)
 
-    dw = dw_plus - dw_minus
+    w += dw_plus - dw_minus
 
-    return dw, STDPState(x=x_decayed, y=y_decayed)
+    # Bound checking
+    w = p_stdp.bounding_func(w)
 
+    return (z_post, state_post, w, state_stdp)
 
-def conv2d_hard_multiplicative_stdp_step(
-    w: torch.Tensor,
+
+# %% Conv2D stepper
+def lif_conv2d_stdp_step(
```

Comment:
This is really neatly done. I can't help but wonder, though, if it can be generalised. If the approach depends on the neuron (`N`) and convolution (`C`), wouldn't that require `N * C` implementations?

Isn't there a more general approach where the same steps are taken, but in a general sense? One naïve suggestion (I'm not an STDP expert :-) ) could be to generalise the stdp part into a forward step and an update step. Something like this:
```python
def stdp_step(x: torch.Tensor, forward: Callable[...], update: Callable[...], p: STDPParameters, cell_state: SomeCellState):
    z, cell_state_post = forward(x, cell_state)
    new_state, p = update(x, z, p, cell_state, cell_state_post)
    return (z, new_state, p)
```
Could this make sense?

I imagine this is related to #94, since improved synapse abstractions would make it easier to operate on this.

Reply:
I would say that the different neuron models are less of a problem than the `Linear` <-> `Conv2d`, especially given that things like #94 could make neuron models more alike. The problem separating `Linear` and `Conv2d` however is the receptive field of each neuron: while this is basically all previous neurons for the former, it is the convolutional filter for the latter, which requires some special ops. I guess it's OK if we have 2 implementations (linear and convolutional STDP) that work for all neurons

Reply:
Agreed. I thought quite a bit about this. Essentially linear and conv2d are two fundamentally different ways of calculating. 

Btw, I think we should also have an explicit conv2d spiking layer. Eventually the weight updates can simply be merged within the update step of the layers. It will take some doing though.

Reply:
Aha, @Jegp does make an excellent point though. Separating the two (stepping the layer and stepping the stdp) would make the implementation more general. I guess that was the idea behind @Huizerd 's initial implementation too.

Reply:
Option#1  Removing the code for the lif stepper. This would mean the user needs to make sure they run the state updates before passing them on to the STDP stepper. It makes the library less user-friendly imo.

Option#2 A clean way to implement an integrated stepper would be if there were an FFState superclass (it can be an abstract class) which all layer-types inherit from. In C++ this is where one would specify the minimum required structure for the layer class, allowing the user to implement their own, library-compatible neurone models. I am sure there's an equivalent or better idiom for this in python!

I will implement option#1 for now.

Reply:
Option#1  Removing the code for the lif stepper. This would mean the user needs to make sure they run the state updates before passing them on to the STDP stepper. It makes the library less user-friendly imo.

Option#2 A clean way to implement an integrated stepper would be if there were an FFState superclass (it can be an abstract class) which all layer-types inherit from. In C++ this is where one would specify the minimum required structure for the layer class, allowing the user to implement their own, library-compatible neurone models. I am sure there's an equivalent or better idiom for this in python!

I will implement option#1 for now.